perun-dispatcher/src/test/java/cz/metacentrum/perun/dispatcher/unit/TaskSchedulerTest.java

```diff
@@ -47,4 +81,134 @@ public void sendToEngineTest() {
+	}
+
+	@Test
+	public void sourceUpdatedRunTest() throws InterruptedException, ExecutionException, TimeoutException {
+		//SchedulingPool schedulingPool = new SchedulingPoolImpl();
+		simpleSpy.setTask(simpleFutureTask);
+		simpleSpy.setSchedulingPool(schedulingPool);
+		Long timeLimit = 100L;
+		Task[] tasks = simpleSetup(timeLimit, schedulingPool);
+		Task testTask1 = tasks[0], testTask2 = tasks[1];
+		testTask2.setSourceUpdated(true);
+		schedulingPool.scheduleTask(testTask2, 4);
+		schedulingPool.scheduleTask(testTask1, 4);
+		Thread.sleep(timeLimit / 100);
+		schedulingPool.scheduleTask(testTask1, 4);
+		Thread.sleep((timeLimit / 100) * 8);
+		schedulingPool.scheduleTask(testTask1, 4);
+
+		simpleFutureTask.run();
+		assertFalse(simpleSpy.testFailed);
+	}
+
+	@Test
+	public void sourceUpdatedRecurrenceRunTest() throws InterruptedException, ExecutionException, TimeoutException {
+		//SchedulingPool schedulingPool = new SchedulingPoolRecurrenceSpy();
+		simpleSpy.setTask(recurrenceFutureTask);
+		recurrenceSpy.setSchedulingPool(schedulingPool);
+		Long timeLimit = 100L;
+		Task[] tasks = simpleSetup(timeLimit, schedulingPool);
+		Task testTask1 = tasks[0];
+		schedulingPool.scheduleTask(testTask1, 2);
+
+		recurrenceFutureTask.run();
+		assertFalse(recurrenceSpy.testFailed);
+	}
+
+	private Task[] simpleSetup(Long timeLimit, SchedulingPool schedulingPool) {
+		Properties properties = ((SchedulingPoolImpl)schedulingPool).getDispatcherProperties();
+		if(properties == null) {
+			properties = dispatcherProperties;
+		}
+		properties.setProperty("dispatcher.task.delay.time", timeLimit.toString());
+		properties.setProperty("dispatcher.task.delay.count", "1");
+		((SchedulingPoolImpl)schedulingPool).setDispatcherProperties(properties);
+		Task testTask1 = new Task();
+		testTask1.setService(service1);
+		testTask1.setFacility(facility1);
+		testTask1.setId(0);
+		testTask1.setStatus(TaskStatus.WAITING);
+		Task testTask2 = new Task();
+		testTask2.setId(2);
+		testTask2.setService(service2);
+		testTask2.setFacility(facility1);
+		testTask2.setStatus(TaskStatus.WAITING);
+		
+		return new Task[]{testTask1, testTask2};
+	}
+
+	private abstract class AbstractTaskSchedulerSpy extends TaskScheduler {
+		@Override
+		protected void initPerunSession() throws InternalErrorException {
+		}
+	}
+
+	/* WHY????
+	private class SchedulingPoolRecurrenceSpy extends SchedulingPoolImpl {
+		@Override
+		public void scheduleTask(Task task, int delayCount) {
+			super.scheduleTask(task, delayCount);
+		}
+	}
+	*/
+
+	private class SimpleTaskSchedulerSpy extends AbstractTaskSchedulerSpy {
+		int scheduledCounter = 0;
+		int scheduleLimit;
+		boolean testFailed = true;
+		FutureTask task;
+
+		public SimpleTaskSchedulerSpy(int scheduleLimit) {
+			this.scheduleLimit = scheduleLimit;
+		}
+
+		public void setTask(FutureTask task) {
+			this.task = task;
+		}
+
+		@Override
+		protected TaskScheduled sendToEngine(Task task) {
+			if (task.getId() == 2) {
```

Comment:
I'm not sure why Task 2 should always fail. Because when I debugged ``sourceUpdatedRunTest()`` test, queue contained 4 tasks (1xTask2 / 3xTask1) and even if you set "sourceUpdated=true" to Task2, by the time TaskScheduler process resolves it, it is switched correctly to "sourceUpdated=false". So you should probably revisit value of ``timeLimit`` (how long should task be waiting for actions like source updated or forced) if you are really trying to catch the case, that WAITING task was incorrectly send as PLANNED to engine. Then you should probably check, that Task in this method doesn't have "sourceUpdated=true" instead of relaying on its ID.

Also repeating adding of Task1 in a test after some time doesn't make sense to me. Maybe it would matter, if you test, that Task1 was switched to PLANNED after first run, but there is no such logic in a test. I'm not sure know, if there should be always only single Task in a queue (by equals) or not. But based on Test, there are multiple TaskFutures for each call ``schedulingPool.scheduleTask(testTask1, 4)`` from the test. Maybee its test-specific and it works correctly in real code - check before adding? - or this is intentional.

I removed calls ``schedulingPool.scheduleTask(testTask1, 4);`` on line 125 and 127 and with the rest of changes in my comments test pass and doesn't stuck. But I'm not sure it really tests what we want to test. There is sadly no explicit declaration, what this test want to to test. 

By TaskScheduler implementation, Tasks like:
``
(task.isSourceUpdated() && schedule.getDelayCount() > 0 && !task.isPropagationForced())
``
are re-scheduled with lowered counter, but it doesn't affect other TaskFutures in a queue and also it doesn't call "sendToEngine()" which is overridden in this test class. So again we should think about, what we want to test.

Reply:
I have no idea what this test should do, I did not write it.

Reply:
OK, can then please make changes I wrote about ? It will stop the test from failing and we can iterate on that later. Thank you.

Reply:
Tried to apply your comments, but please, if you have some changes in mind, I'd prefer patch - sometimes the comments are not clear to me. 

The dispatcher tests are failing for me with 

2018-03-16 11:37:04,392 ERROR TransactionAspectSupport.java:539 - Application exception overridden by rollback exception
cz.metacentrum.perun.core.api.exceptions.InternalErrorException: Error 1622e6269ff: PreparedStatementCallback; SQL [select  services.id as services_id, services.name as services_name, services.description as services_description, services.delay as services_delay, services.recurrence as services_recurrence, services.enabled as services_enabled, services.script as services_script, services.created_at as services_created_at, services.created_by as services_created_by, services.modified_by as services_modified_by, services.modified_at as services_modified_at, services.created_by_uid as services_created_by_uid, services.modified_by_uid as services_modified_by_uid from services where id=?]; transaction rollback: serialization failure; nested exception is java.sql.SQLTransactionRollbackException: transaction rollback: serialization failure

Reply:
Ok, noted.

Serialization failure is caused by returning ``return TaskScheduled.ERROR;`` from the ``sendToEngine()`` since then ``TaskScheduler`` continues with its method ``run()`` and re-schedule the Task by inserting it to the pool again (just like test do before calling run()). But it seems, that since this is done outside of test transaction, it causes fail (service is not found etc.).

Just fix condition in ``sendToEngine()`` to something like this:

``(task.isSourceUpdated() || schedule.getDelayCount() > 0) && !task.isPropagationForced()``

Like this it will test TaskScheduler inner condition, that not forced tasks with source updated or waiting in delay are not sent to engine.

Reply:
Condition fixed, serialization failure disappeared, tests are still failing.

Reply:
My bad, I copy/pasted and modified production code. Just check:

``task.isSourceUpdated() && !task.isPropagationForced()``

since ``schedule`` variable is not present inside ``sendToEngine()`` method but in above logic.